### ISSUE TITLE
Assorted problems in calling the wrong wrapper for reachesIntoOuterContext 

### ISSUE THREAD

**Issue by kaby76** [antlr/antlr4#4050]
This is a serious bug in the CSharp runtime, ParserATNSimulator.cs.

[reachesIntoOuterContext](https://github.com/antlr/antlr4/blob/76fa05c21b12b96a6c12a0a82e611ed9d87d5af4/runtime/CSharp/src/Atn/ATNConfig.cs#L67) is an integer that tracks the depth of how far we dip into the outer context. This field must be interpreted with the bit map mask [SUPPRESS_PRECEDENCE_FILTER](https://github.com/antlr/antlr4/blob/76fa05c21b12b96a6c12a0a82e611ed9d87d5af4/runtime/CSharp/src/Atn/ATNConfig.cs#L30), and *careful* attention placed on whether to access the field raw or with the bit map mask.

In Java, [this code](https://github.com/antlr/antlr4/blob/76fa05c21b12b96a6c12a0a82e611ed9d87d5af4/runtime/Java/src/org/antlr/v4/runtime/atn/ParserATNSimulator.java#L1497) accesses the field raw. 

In CSharp, [the field is accessed through a method that applies the bit map mask](https://github.com/antlr/antlr4/blob/76fa05c21b12b96a6c12a0a82e611ed9d87d5af4/runtime/CSharp/src/Atn/ParserATNSimulator.cs#L1572). This is a serious problem in that it causes ATN parser trace divergence.

I did a cursory check on the other targets, and I am not confident the field access is done correctly across targets.

The fix is the change ParserATNInterpreter.cs:
```
diff --git a/runtime/CSharp/src/Atn/ParserATNSimulator.cs b/runtime/CSharp/src/Atn/ParserATNSimulator.cs
index 4a7a6a3d5..8463bfcc3 100644
--- a/runtime/CSharp/src/Atn/ParserATNSimulator.cs
+++ b/runtime/CSharp/src/Atn/ParserATNSimulator.cs
@@ -1579,7 +1579,7 @@ namespace Antlr4.Runtime.Atn
 						// This assignment also propagates the
 						// isPrecedenceFilterSuppressed() value to the new
 						// configuration.
-						c.reachesIntoOuterContext = config.OuterContextDepth;
+                                                c.reachesIntoOuterContext = config.reachesIntoOuterContext;
 						ClosureCheckingStopState(c, configSet, closureBusy, collectPredicates,
 												 fullCtx, depth - 1, treatEofAsEpsilon);
 					}
```

(Note, the source code should **really** not be using tabs. Tab expansion is editor specific.)
```

**Comment by ericvergnaud** [comment]
You're right, and git blame doesn't tell us much. Looks like this is a typo from the very early days... Will you be submitting a PR across all targets ?

**Comment by kaby76** [comment]
> You're right, and git blame doesn't tell us much. Looks like this is a typo from the very early days... Will you be submitting a PR across all targets ?

I'm working on the PHP target with a number of critical changes. So, I will prefer not to work on this issue, too much to do. Thanks. Again, so glad to see ATN tracing added @parrt !

**Comment by parrt** [comment]
Thanks for the heads up!
